packages/clay-list/src/List.tsx

```diff
@@ -41,7 +49,19 @@ const ClayList: React.FunctionComponent<IProps> & {
 				'show-quick-actions-on-hover': showQuickActionsOnHover,
 			})}
 		>
-			{children}
+			{children &&
+				React.Children.map(children, (child) => {
+					warning(
+						typeof child.type === 'string' ||
+							// @ts-ignore
+							(child.type.displayName !== 'ClayListItem' &&
+								// @ts-ignore
+								child.type.displayName !== 'ClayListHeader'),
+						'ClayList: Direct descendant of ClayList must be either ClayList.Item or ClayList.Header.'
```

Comment:
I'm just worried about one thing, it may happen that some people pass the primitive `li` here, do we want to ignore it from the warning? in theory, we should but this could be a way to force the use of `Item` and `Header` although both cover the edge cases. What do you think, does it make sense?

Reply:
Thats probably fine to check for `li` and allow it since that is semantic. Although, now this makes me re-think even adding this because its possible they could create their own custom component that renders an `li` and then that would fail this check since the displayName would be something other than the clay components.

Unless someone has another idea of how to enforce this via ts, I'm thinking we scrape this and add a note in the docs instead. cc @wincent

Reply:
This is why [I said](https://github.com/liferay/clay/pull/3278#discussion_r431653472):

> Makes me wonder whether there should be a check (only when `process.env.NODE_ENV === 'development'`) that looks at the children and makes sure there are no non-list children... (or **at least no children that are definitely the wrong type of Clay component**).

(Emphasis added). That is to say, we can warn if you use the wrong Clay component (ie. `displayName` is `ClaySomething` but not `ClayListItem` or `ClayListHeader`), so it would catch obvious errors, and remain silent about non-obvious ones. ie. some protection being better than none.

> add a note in the docs instead

Never a bad idea.

I'm fine with doing the check or not doing it.

Reply:
@wincent ah I see, so you are thinking checking against `Clay*` but allowing ListItem and ListHeader? I think that makes sense to me, especially on a library level.

Ill add that plus the docs and update here